emma-language/src/main/scala/org/emmalanguage/compiler/lang/comprehension/Combination.scala

```diff
@@ -442,29 +451,33 @@ private[comprehension] trait Combination extends Common {
       } yield {
         // It can happen that the key functions have differing return types (e.g. Long and Int).
         // In this case, we add casts before the return expr of the key functions to the weakLub.
-        def maybeCast(tree: u.Tree) = if (kxExpr.tpe =:= kyExpr.tpe) tree else tree match {
+        val kLub = api.Type.weakLub(Seq(kxExpr.tpe, kyExpr.tpe))
+        def maybeCast(tree: u.Tree) = if (tree.tpe <:< kLub) tree else tree match {
```

Comment:
Could you explain changing `kxExpr.tpe =:= kyExpr.tpe` to `tree.tpe <:< kLub`? I'm not sure I correctly understand what `<:<` means, but if it means that a cast won't be inserted if the key expression is returning a subtype of `kLub`, then I think this change will break things.

Note, that the return type of the two key functions that we create must be _exactly_ the same, to be able to pass them to `equiJoin`.

For example, look at CombinationSpec:282, and imagine what would happen if these were not primitive types, but some classes, and thus `<:<` would say that the cast is not needed, and then the return type of `kx` would be bad when passing it to `equiJoin`.